### ISSUE TITLE
Keep cookie connect persistent

### ISSUE THREAD

**Issue by lesonunique** [signalpoint/DrupalGap#643]
Hello,

I use system_connect() for load a user connected when the application lauch. It work but sometime, the user is not load.

Typically , this happens when you close the application completely ( by dragging the multitasking ).
![screenshot_2015-10-07-17-10-59](https://cloud.githubusercontent.com/assets/12293195/10342105/cbbecb66-6d17-11e5-83c9-a0143960ff04.png)

I see the cookie of the session in debug console and it is no longer present when it no longer works.
![sans titre-1](https://cloud.githubusercontent.com/assets/12293195/10342169/15c67e16-6d18-11e5-87a8-73f7dcd1e10d.png)

It's possible to keep the cookies persistent ? I use Cordova CLI-5.3.0

Thank you in advance.

**Comment by signalpoint** [comment]
@lesonunique I believe this is why many apps go with something like OAuth, which is used for persistent sessions (AFAIK). That being said, a quick Google Search reveals many articles discuss PhoneGap and persistent cookies. I believe this issue lies more up in jDrupal and/or PhoneGap land (with a little Drupal Services tossed in).

**Comment by lesonunique** [comment]
I tried to stock in localstorage the SESSID and the Session_name (data present in the cookie) and i tried to put it in a cookie for always have it.

The data are always stock in the data correctly but I can't stock the cookie because Android 5.0 block the local cookie. 

I find this solution for the same problem, but it change nothing :
https://mytechead.wordpress.com/2014/03/04/android-enable-local-cookies-in-phonegap-webview/

There he has another solution to allow local cookie?

Or is it possible to execute the requests "?q=services/session/token" and "?q=drupalgap/user/1321.json" (or other ?) with SESSID and Session_name data (and therefore no longer use the cookie but only the localstorage in cordova app) ?

Thank you.

Vincent

**Comment by signalpoint** [comment]
@lesonunique Yes, I believe that if the session id and name are set aside somewhere in local storage (this may be a security issue though, needs some research), then those values can be passed along with the request in place of the cookie (which it sounds like the phonegap/webview cookie is expired by the native OS). This feature should live up in jDrupal for sure.

**Comment by lesonunique** [comment]
Thanks but I don't find where put the data in the request ? Any idea ?

**Comment by signalpoint** [comment]
@lesonunique You'll want to become familiar with this file if you're going to append new things to the service call: https://github.com/easystreet3/jDrupal/blob/7.x-1.x/src/services/services.js

Also, to set aside the session id and name, you'll want to do that in the success callback for the system connect resource: https://github.com/easystreet3/jDrupal/blob/7.x-1.x/src/services/services.system.js

**Comment by lesonunique** [comment]
Yes I searched a lot in this file. 

I see how retrieves a token with the username and password . 
But I do not understand where make the request with the SESSID and session_name. 

How system_connect retrieves the cookie and where? it will be here that I have to put the data stored locally.

I searched a lot and I still have difficulty understanding the entire operation of the script.

Thank you.

Vincent

**Comment by signalpoint** [comment]
@lesonunique I believe the session id and name need to be appended to the request as an HTTP header.

As for the system connect and the cookie, I'm fairly certain that the cookie is set by Drupal into the user's browser (phonegap browser, web app browser, ect). That being said, I think the system connect result contains the session id and name (otherwise it'll have to be extracted from the cookie if possible, or retrieve elsewhere. where? I'm not sure), so during the success of the system connect, those values should be set aside in local storage, so they can be retrieved later during all service calls.

Don't feel bad for the difficulty, this is indeed a complex issue.

**Comment by lesonunique** [comment]
Sure, I could get SESSID and session_name and easily to store in localStorage. My problem is mainly to reuse this information to launch the application to load an already logged in user.

I'll try to look a little more, so it would be with the parameters in system_connect ?

**Comment by signalpoint** [comment]
@lesonunique Yes, I think if the session id and name are appended as a single header to the system connect call (and maybe others), I think as a single string with some kind of delimiter separating the two, I've seen some docs floating around on this. I'm not sure what the name of the header value would be though (or if this is the right approach, just going by memory hear).

Then basically when we would preprocess calls to system connect (and others?) in https://github.com/easystreet3/jDrupal/blob/7.x-1.x/src/services/services.system.js, by attaching the session id/name header (or whatever needs to be attached). This will be interesting though, because we won't want to do this if a cookie is present, so if we can detect a cookie too, that'd be good.

**Comment by lesonunique** [comment]
Hello,

Finally I created a php page to which I execute a request with the SESSID and session_name (stored in local storage) and execute with php page a setcookie() for create the cookie at start to load user.

It works generaly, but occasionally the loading of the user fails. I feel that it happens when I disconnect my user, reconnects it, close and relauch the application.

I have the error "401 Unauthorized: CSRF validation failed". But if I close and reopen the application without doing anything, the reconnection work.

Do you have an idea of solving the problem? Can be an action that is made when a user sign out that disrupts this?

What is the reason for the error "401 Unauthorized: CSRF validation failed"?

Thank you in advance.

Vincent

**Comment by signalpoint** [comment]
@lesonunique The infamous `401 Unauthorized: CSRF validation failed` error means one of the following:
- the X-CSRF-Token header is missing from the request (value available here `?q=services/session/token`)
- the X-CSRF-Token header value is invalid

For the latter reason, I'm not 100% sure why that may happen, but I'm fairly certain any token values are related to a particular session id/name, so over time I could see the token becoming invalid, or upon system connect a new token is retrieved and it doesn't match with the original token.

I wouldn't recommend storing the X-CSRF-Token in local storage, that too seems like a security risk.
